src/codestudio.js

```diff
@@ -0,0 +1,100 @@
+let redact;
+
+const CODESTUDIO_RE = /^(\[code-studio\s*)(\d+)?-?(\d+)?\]/;
+const CODESTUDIO = "codestudio";
+
+/**
+ * This plugin renders the special "codestudio" syntax used by
+ * CurriculumBuilder for the code studio pullthrough.
+ *
+ * Specifically, given syntax like `[code-studio]`, optionally with range
+ * parameters like `[code-studio 1]` or `[code-studio 3-5]`, it renders a div
+ * with the class "stage-guide" and the optional parameters included as data
+ * attributes.
```

Comment:
In addition to exactly what DOM is rendered, I think it'd be helpful to mention what we eventually do with this empty, annotated div.  My assumption is that some JavaScript picks this up and swaps in a "stage guide."  It looks like there are a few separate implementations of this in the curriculumbuilder repo ([here's one](https://github.com/code-dot-org/curriculumbuilder/blob/69b1f1ec2927a98bbf14248df393b8bd52883414/static/js/lesson.js#L63)) so finding a canonical one to point to might be difficult; still, some sense of what you'd expect an end user to eventually see in the place where you used this syntax, and how the optional parameters affect that, would be helpful.  Maybe referencing the role of `.stage_guide_hidden` would be relevant?

Reply:
I've gone back and forth on this a lot, and I can't decide whether or not it's the responsibility of the markdown processor to comment on how the generated HTML is intended to be used in a specific context. I feel like the ideal design of a markdown processor is one that is implementation agnostic; this specific component obviously violates that ideal, but is it better for a violation to be a complete violation, or should it be minimal?

Of course, it's all exacerbated by the fact that - as you point out - we don't even have a canonical implementation to reference. Perhaps a comment like "here's an example of one of the ways we use this HTML, but there are also others"?

Reply:
I agree that the implementation of this plugin should be ignorant of its eventual use.  In its documentation I have no qualms about additional context on the original intended use of the plugin.  I'd argue that one probably _shouldn't_ use this very specific plugin without some additional system that uses the generated empty divs.  I'll make a stronger assertion: Even though you've isolated the implementation, I still think of this plugin called "code studio" as more like application code than library code, inextricably tied to the global context of our particular product, and I think it would be helpful for it to document its place within that product. It does seem like a good idea to qualify any such description as you suggested, with "one of the ways we use this" or "the original way we used this" so that the comment doesn't go out-of-date as quickly.

But very curious how much this breaks with your frame of reference!  Let me know what you think.